graylog2-server/src/main/java/org/graylog2/database/pagination/DefaultMongoPaginationHelper.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog2.database.pagination;
+
+import com.google.common.primitives.Ints;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.Collation;
+import com.mongodb.client.model.Sorts;
+import org.bson.conversions.Bson;
+import org.graylog2.database.MongoEntity;
+import org.graylog2.database.PaginatedList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+import static org.graylog2.database.utils.MongoUtils.stream;
+
+public class DefaultMongoPaginationHelper<T extends MongoEntity> implements MongoPaginationHelper<T> {
```

Comment:
Let's document that this implementation is immutable and can be re-used.

Reply:
done